src/passes/TranslateEH.cpp

```diff
@@ -0,0 +1,823 @@
+    }
+    auto* tryTable = builder.makeTryTable(
+      curr->body, catchTags, catchDests, catchRefs, curr->type);
+
+    // If we don't have any catches, we don't need to do more.
+    if (curr->catchBodies.empty()) { // catch-less try
+      replaceCurrent(tryTable);
+      return;
+    }
+
+    // Now we convert catch bodies.
+    //
+    // For simplicity, let's assume all tags have the none type and there are no
+    // rethrows (which have converted to throw_refs at this point) and the try's
+    // type is none as well. Then
+    //
+    // (try
+    //   (catch $e
+    //     (catch_body)
+    //   )
+    //   (catch_all
+    //     (catch_all_body)
+    //   )
+    // )
+    //
+    // becomes
+    //
+    // (block $outer
+    //   (block $catch_all
+    //     (block $catch
+    //       (try_table (catch $e $catch) (catch_all $catch_all)
+    //         ...
+    //       )
+    //       (br $outer)
+    //     )
+    //     (catch_body)
+    //     (br $outer)
+    //   )
+    //   (catch_all_body) ;; We don't need (br $outer) for the last catch
+    // )
+    //
+    // Here (block $outer) has been already created and given as an argument to
+    // this function. This is the outer block that will contain the whole
+    // structure.
+    //
+    // If there are more catch clauses, there will be more layers of blocks,
+    // catch bodies, and (br $outer)s.
+    //
+    // If try's type is concrete, (br $outer)s will contain try_table and catch
+    // bodies as values:
+    //
+    // (block $outer
+    //   (block $catch_all
+    //     (block $catch
+    //       (br $outer
+    //         (try_table (result sometype)
+    //                    (catch $e $catch) (catch_all $catch_all)
+    //           ...
+    //         )
+    //       )
+    //     )
+    //     (br $outer
+    //       (catch_body)
+    //     )
+    //   )
+    //   (catch_all_body) ;; We don't need (br $outer) for the last catch
+    // )
+    //
+    // ---
+    //
+    // When a tag has a concrete type, we assign it to a scratch local of that
+    // type, and (pop tagtype) would have been already converted to
+    // (local.get $scratch) in visitPop(). These scratch locals have extremely
+    // short lifetimes; basically they are read right after they are written, so
+    // we can reuse them throughout the function.
+    //
+    // So, when there is no rethrows (throw_refs), and if we assume that try's
+    // type is none for simplicity,
+    // (try
+    //   ...
+    //   (catch $e-i32 ;; concrete type
+    //     (use_inst
+    //       (pop i32)
+    //     )
+    //   )
+    // )
+    //
+    // becomes
+    // (block $outer
+    //   (local.set $scratch-i32
+    //     (block $catch (result tagtype)
+    //       (try_table (catch $e-i32 $catch)
+    //         ...
+    //       )
+    //       (br $outer)
+    //     )
+    //   )
+    //   (use_inst
+    //     (local.get $scratch-i32)
+    //   )
```

Comment:
In case of `catch`, we can, because `pop` is always the first instruction in the catch body in the execution order. So if there's a single `catch`, it can be like
```wast
(block $outer
  (use_inst
    (block $l0 (result i32)
      (try_table (catch $l0 $tag)
        ...
      )
      (br $outer)
    )
  )
)
```

But in case of `catch_all` (used for destructors in C++), it's hard, because there would be another instruction (calls to a destructor) before a `throw_ref`, so `throw_ref` is not the first instruction within a `catch_all` body. For example, in case of a single `catch_all`,
```wast
(block $outer
  (call $destructor) ;; We can't do this!
  (throw_ref
    (block $l0 (result exnref)
      (try_table (catch $l0 $tag)
        ...
      )
      (br $outer)
    )
  )
)
```
We can't do this because we can't bypass the `(call $destructor)`.

And really the only thing that matters to the code size for C++ programs is destructor calls with `try`-`catch_all`, because there are so many of them. The number of `try`-`catch`, which is generated from C++ `try`-`catch` clauses, is so small compared to that of `try`-`catch_all`. (J2Wasm currently doesn't even use `catch_all` and a very small number of `catch`es, so we don't need to worry about it)

So if we can't transform `try`-`catch_all`, I think keeping the current transformation is simpler, because we can apply the same way of transformation to all `catch` variants. If we do one thing for `catch` and something else for `catch_all_ref` that will be more complicated and that doesn't decrease code size much anyway. Also after implementing Binaryen support for optimization pipeline, I think it can take care of those `catch`es later.